akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -1964,148 +1965,181 @@ private[stream] object Collect {
-          setHandler(in, new InHandler {
-            override def onPush(): Unit = throw new IllegalStateException("LazyFlow received push while waiting for flowFactory to complete.")
-            override def onUpstreamFinish(): Unit = gotCompletionEvent()
-            override def onUpstreamFailure(ex: Throwable): Unit = failure(ex)
-          })
+          case Failure(e) ⇒
+            matPromise.failure(e)
+            failStage(e)
+        }
+        try {
+          flowFactory(element).onComplete(cb.invoke)(ExecutionContexts.sameThreadExecutionContext)
         } catch {
-          case NonFatal(e) ⇒ decider(e) match {
-            case Supervision.Stop ⇒ failure(e)
-            case _                ⇒ pull(in)
-          }
+          case NonFatal(e) ⇒
+            matPromise.failure(e)
+            failStage(e)
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        // ignore onUpstreamFinish while the stage is switching
+        if (!switching) {
+          matPromise.success(None)
+          super.onUpstreamFinish()
         }
       }
 
+      override def onUpstreamFailure(ex: Throwable): Unit = {
+        matPromise.failure(ex)
+        super.onUpstreamFailure(ex)
+      }
+
+      override def onDownstreamFinish(): Unit = {
+        matPromise.success(None)
+        super.onDownstreamFinish()
+      }
+
       override def onPull(): Unit = {
         pull(in)
-        subSink.pull()
+      }
 
-        setHandler(out, new OutHandler {
-          override def onPull(): Unit = {
-            subSink.pull()
-          }
+      setHandler(in, this)
+      setHandler(out, this)
 
-          override def onDownstreamFinish(): Unit = {
-            subSink.cancel()
+      private def switchTo(flow: Flow[I, O, M], firstElement: I): M = {
+
+        var firstElementPushed = false
+
+        val subInlet = new SubSinkInlet[O]("LazyFlowSubSink")
+        val subOutlet = new SubSourceOutlet[I]("LazyFlowSubSource")
+
+        val matVal = Source.fromGraph(subOutlet.source)
+          .viaMat(flow)(Keep.right)
+          .toMat(subInlet.sink)(Keep.left)
+          .run()(interpreter.subFusingMaterializer)
+
+        // The lazily materialized flow may be constructed from a sink and a source. Therefore termination
+        // signals (completion, cancellation, and errors) are not guaranteed to pass through the flow. This
+        // means that this stage must not be completed as soon as one side of the flow is finished.
+        //
+        // Invariant: isClosed(out) == subInlet.isClosed after each event because termination signals (i.e.
+        // completion, cancellation, and failure) between these two ports are always forwarded.
+        //
+        // However, isClosed(in) and subOutlet.isClosed may be different. This happens if upstream completes before
+        // the cached element was pushed.
+        def maybeCompleteStage(): Unit = {
+          if (isClosed(in) && subInlet.isClosed && isClosed(out)) {
```

Comment:
Maybe a test case for the scenario from the comment above?

Reply:
I removed the supervision strategy from the tests.

I changed the test case `complete normally when upstream is complete` into a test case that checks `complete normally when upstream completes BEFORE the stage has switched to the inner flow`. The check is implemented by delaying the completion of the flowFactory future.

I added another test `complete normally when upstream completes AFTER the stage has switched to the inner flow` that checks if completion propagates properly after the stage has switched to the inner flow. That test is implemented by delaying the completion until the first element has passed throught the flow.